beacon-chain/sync/validate_aggregate_proof.go

```diff
@@ -0,0 +1,142 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/go-ssz"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/blocks"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/bls"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/roughtime"
+	"go.opencensus.io/trace"
+)
+
+// validateAggregateAndProof verifies the aggregated signature and the selection proof is valid before forwarding to the
+// network and downstream services.
+func (r *RegularSync) validateAggregateAndProof(ctx context.Context, msg proto.Message, p p2p.Broadcaster, fromSelf bool) (bool, error) {
+	ctx, span := trace.StartSpan(ctx, "sync.validateAggregateAndProof")
+	defer span.End()
+
+	// To process the following it requires the recent blocks to be present in the database, so we'll skip
+	// validating or processing aggregated attestations until fully synced.
+	if r.initialSync.Syncing() {
+		return false, nil
+	}
+
+	m, ok := msg.(*pb.AggregateAndProof)
+	if !ok {
+		return false, nil
+	}
+
+	attSlot := m.Aggregate.Data.Slot
+
+	// Verify aggregate attestation has not already been seen via aggregate gossip, within a block, or through the creation locally.
+	// TODO(3835): Blocked by operation pool redesign
+
+	// Verify the block being voted for passes validation. The block should have passed validation if it's in the DB.
+	if !r.db.HasBlock(ctx, bytesutil.ToBytes32(m.Aggregate.Data.BeaconBlockRoot)) {
+		return false, errPointsToBlockNotInDatabase
+	}
+
+	s, err := r.chain.HeadState(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	// Verify attestation slot is within the last ATTESTATION_PROPAGATION_SLOT_RANGE slots.
+	currentSlot := (uint64(roughtime.Now().Unix()) - s.GenesisTime) / params.BeaconConfig().SecondsPerSlot
+	if attSlot > currentSlot || currentSlot > attSlot+params.BeaconConfig().AttestationPropagationSlotRange {
```

Comment:
Can you add a test for this? I commented it out and all of your tests passed.

Reply:
Nice catch. Just added `TestValidateAggregateAndProof_NotWithinSlotRange`